File: src/Firegrid.Core/StateDiff.fs
namespace Firegrid.Core

open Fable.Core
open Fable.Core.JsInterop

/// Minimal JSON Patch (RFC 6902) operation. Port of `state-diff.ts`'s
/// `Operation` union — `replace`/`add` carry a value, `remove` does not.
type Operation =
    | Replace of path: string * value: obj
    | Add of path: string * value: obj
    | Remove of path: string

[<RequireQualifiedAccess; CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
module Operation =

    /// JS-object form `{ op, path, value? }` for the wire / log payload.
    let toObj (op: Operation) : obj =
        match op with
        | Replace(path, value) -> createObj [ "op" ==> "replace"; "path" ==> path; "value" ==> value ]
        | Add(path, value) -> createObj [ "op" ==> "add"; "path" ==> path; "value" ==> value ]
        | Remove path -> createObj [ "op" ==> "remove"; "path" ==> path ]

[<RequireQualifiedAccess>]
module StateDiff =

    /// `isObject` — value is non-null and `typeof === 'object'`.
    let isObject (value: obj) : bool =
        not (isNull value) && CoreSdk.isTypeofObject value

    /// `escapeJsonPointer` — escape `~` then `/` per RFC 6901.
    let escapeJsonPointer (segment: string) : string =
        segment.Replace("~", "~0").Replace("/", "~1")

    /// `normalizeValue` — coerce `undefined` to `null` recursively so the
    /// serialized op is always well-formed RFC 6902.
    let rec normalizeValue (value: obj) : obj =
        if CoreSdk.isUndefined value then
            box null
        elif CoreSdk.isArray value then
            let arr = unbox<obj[]> value
            box (Array.map normalizeValue arr)
        elif isObject value then
            let out = createObj []

            for (k, v) in CoreSdk.objectEntries value do
                CoreSdk.setProp out k (normalizeValue v)

            out
        else
            value

    /// `snapshotState` — deep clone a state object via `structuredClone`.
    let snapshotState<'T> (state: 'T) : 'T = CoreSdk.structuredClone state

    let rec private diff (prev: obj) (next: obj) (path: string) : Operation list =
        if CoreSdk.objectIs prev next then
            []
        else
            let prevIsObj = isObject prev
            let nextIsObj = isObject next

            // One is a primitive (or null), or array-ness disagrees — replace node.
            if not prevIsObj
               || not nextIsObj
               || (CoreSdk.isArray prev <> CoreSdk.isArray next) then
                [ Replace((if path = "" then "" else path), normalizeValue next) ]
            elif CoreSdk.isArray prev && CoreSdk.isArray next then
                let prevLen = CoreSdk.arrayLength prev
                let nextLen = CoreSdk.arrayLength next

                if prevLen <> nextLen then
                    [ Replace((if path = "" then "" else path), normalizeValue next) ]
                else
                    [ for i in 0 .. prevLen - 1 do
                          yield!
                              diff
                                  (CoreSdk.arrayItem<obj> prev i)
                                  (CoreSdk.arrayItem<obj> next i)
                                  (sprintf "%s/%d" path i) ]
            else
                // Both plain objects.
                let prevKeys = CoreSdk.objectKeys prev
                let nextKeys = CoreSdk.objectKeys next

                let allKeys =
                    let seen = System.Collections.Generic.HashSet<string>()
                    let ordered = ResizeArray<string>()

                    for k in Array.append prevKeys nextKeys do
                        if seen.Add k then
                            ordered.Add k

                    ordered

                [ for key in allKeys do
                      let subPath = sprintf "%s/%s" path (escapeJsonPointer key)
                      let prevHas = CoreSdk.hasOwnProperty prev key
                      let nextHas = CoreSdk.hasOwnProperty next key

                      if prevHas && nextHas then
                          yield! diff (CoreSdk.prop<obj> prev key) (CoreSdk.prop<obj> next key) subPath
                      elif nextHas then
                          yield Add(subPath, normalizeValue (CoreSdk.prop<obj> next key))
                      else
                          yield Remove subPath ]

    /// `diffState` — produce an RFC 6902 patch from `prev` to `next`.
    let diffState (prev: obj) (next: obj) : Operation list = diff prev next ""

File: packages/core/src/engine/state-diff.ts
// @ts-nocheck -- Vendored TanStack source targets a looser optional-property TypeScript policy.
/* oxlint-disable effect/restricted-syntax -- Vendored TanStack implementation source keeps upstream imperative control flow. */
/**
 * Minimal JSON Patch (RFC 6902) helpers for workflow state trace.
 *
 * Emits the three op kinds the engine needs (replace, add, remove).
 * Clients applying these patches handle the same set. Move/copy/test
 * are intentionally omitted — they're never produced by a forward diff
 * and the spec allows producers to use any subset.
 */

export type Operation =
  | { op: "replace"; path: string; value: unknown }
  | { op: "add"; path: string; value: unknown }
  | { op: "remove"; path: string }

/**
 * Snapshot a state object for later diffing.
 */
export function snapshotState<T>(state: T): T {
  return structuredClone(state)
}

/**
 * Produce an RFC 6902 JSON Patch from `prev` to `next`. Empty array if
 * no changes. Recursively diffs plain objects and arrays; for arrays of
 * different length, emits a single top-level `replace` rather than
 * splice-style ops (simpler wire shape, sufficient for state
 * trace).
 */
export function diffState<T>(prev: T, next: T): Array<Operation> {
  return diff(prev, next, "")
}

function diff(prev: unknown, next: unknown, path: string): Array<Operation> {
  if (Object.is(prev, next)) return []

  const prevIsObj = isObject(prev)
  const nextIsObj = isObject(next)

  // One is a primitive (or null), or types disagree — replace whole node.
  if (!prevIsObj || !nextIsObj || Array.isArray(prev) !== Array.isArray(next)) {
    return [{ op: "replace", path: path || "", value: normalizeValue(next) }]
  }

  if (Array.isArray(prev) && Array.isArray(next)) {
    // Length mismatch → replace the array. Same length → diff element-wise.
    if (prev.length !== next.length) {
      return [{ op: "replace", path: path || "", value: normalizeValue(next) }]
    }
    const ops: Array<Operation> = []
    for (let i = 0; i < prev.length; i++) {
      ops.push(...diff(prev[i], next[i], `${path}/${i}`))
    }
    return ops
  }

  // Both are plain objects.
  const prevObj = prev as Record<string, unknown>
  const nextObj = next as Record<string, unknown>
  const ops: Array<Operation> = []
  const allKeys = new Set([...Object.keys(prevObj), ...Object.keys(nextObj)])

  for (const key of allKeys) {
    const subPath = `${path}/${escapeJsonPointer(key)}`
    const prevHas = Object.prototype.hasOwnProperty.call(prevObj, key)
    const nextHas = Object.prototype.hasOwnProperty.call(nextObj, key)

    if (prevHas && nextHas) {
      ops.push(...diff(prevObj[key], nextObj[key], subPath))
    } else if (nextHas) {
      ops.push({
        op: "add",
        path: subPath,
        value: normalizeValue(nextObj[key])
      })
    } else {
      ops.push({ op: "remove", path: subPath })
    }
  }

  return ops
}

/**
 * Normalize `undefined` to `null` recursively before emitting on the
 * wire. `JSON.stringify` drops `undefined` properties, so emitting
 * `{ op: 'add', path: '/x', value: undefined }` produces the RFC 6902
 * invalid `{"op":"add","path":"/x"}` on the wire — clients applying it
 * then either error or silently write `undefined`. Coerce here so the
 * serialized op is always well-formed.
 */
function normalizeValue(value: unknown): unknown {
  if (value === undefined) return null
  if (Array.isArray(value)) return value.map(normalizeValue)
  if (isObject(value)) {
    const out: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      out[k] = normalizeValue(v)
    }
    return out
  }
  return value
}

function isObject(value: unknown): value is object {
  return value !== null && typeof value === "object"
}

/**
 * Escape `/` and `~` per RFC 6901 (JSON Pointer).
 */
function escapeJsonPointer(segment: string): string {
  return segment.replace(/~/g, "~0").replace(/\//g, "~1")
}
